include/LightGBM/category_encoding_provider.hpp

```diff
@@ -0,0 +1,771 @@
+/*!
+  * Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+  * Licensed under the MIT License. See LICENSE file in the project root for license information.
+  */
+#ifndef LIGHTGBM_CATEGORY_ENCODING_PROVIDER_HPP_
+#define LIGHTGBM_CATEGORY_ENCODING_PROVIDER_HPP_
+
+#include <LightGBM/config.h>
+#include <LightGBM/network.h>
+#include <LightGBM/utils/json11.h>
+#include <LightGBM/utils/text_reader.h>
+#include <LightGBM/utils/threading.h>
+#include <LightGBM/parser_base.h>
+
+#include <algorithm>
+#include <memory>
+#include <random>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+namespace LightGBM {
+
+using json11::Json;
+
+// transform categorical features to encoded numerical values before the bin construction process
+class CategoryEncodingProvider {
+ public:
+  class CatConverter {
```

Comment:
Can we put CatConverter as separate class in separate file to reduce complexity for CategoryEncodingProvider? Looks like we might need to access properties in CategoryEncodingProvider, then we can pass CategoryEncodingProvider to the CatConverter.

Reply:
For every converter we can have separate unit tests?